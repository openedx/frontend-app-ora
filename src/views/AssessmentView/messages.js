import { defineMessages } from '@edx/frontend-platform/i18n';
import { stepNames } from 'constants/index';

const messages = defineMessages({
  [stepNames.self]: {
    defaultMessage: 'Self grading',
    description: 'Self assessment view header text',
    id: 'frontend-app-ora.selfAssessmentView.header',
  },
  [stepNames.peer]: {
    defaultMessage: 'Grade your peers',
    description: 'Peer assessment view header text',
    id: 'frontend-app-ora.peerAssessmentView.header',
  },
  [stepNames.studentTraining]: {
    defaultMessage: 'Practice grading',
    description: 'Student training view header text',
    id: 'frontend-app-ora.studentTrainingView.header',
  },
});

const responseMessages = defineMessages({
  [stepNames.self]: {
    defaultMessage: 'Your response',
    description: 'Self assessment view response header text',
    id: 'frontend-app-ora.selfAssessmentView.responseHeader',
  },
  [stepNames.peer]: {
    defaultMessage: 'Peer response',
    description: 'Peer assessment view response header text',
    id: 'frontend-app-ora.peerAssessmentView.responseHeader',
  },
  [stepNames.studentTraining]: {
    defaultMessage: 'Example response',
    description: 'Student training view response header text',
    id: 'frontend-app-ora.studentTrainingView.responseHeader',
  },
});

export default {
  ...messages,
  responseMessages,
};
