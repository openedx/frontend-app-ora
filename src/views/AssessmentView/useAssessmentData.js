import React from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import {
  useIsORAConfigLoaded,
  useIsPageDataLoaded,
  usePrompts,
  useResponse,
  useSetResponse,
  useResponseData,
} from 'hooks/app';

export const stateKeys = StrictDict({
  initialized: 'initialized',
});

const useAssessmentData = () => {
  const [initialized, setInitialized] = useKeyedState(stateKeys.initialized, false);
  const prompts = usePrompts();
  const response = useResponse();
  const responseData = useResponseData();
  const setResponse = useSetResponse();
  const isLoaded = useIsORAConfigLoaded();
  const isPageDataLoaded = useIsPageDataLoaded();
  React.useEffect(() => {
    if (!initialized && isLoaded && isPageDataLoaded) {
      setResponse(responseData);
      setInitialized(true);
    } else if (initialized && responseData && response !== responseData) {
      setResponse(responseData);
    }
  }, [responseData, initialized, isLoaded, isPageDataLoaded, response, setResponse, setInitialized]);
  return {
    isLoaded,
    response,
    prompts,
  };
};

export default useAssessmentData;
