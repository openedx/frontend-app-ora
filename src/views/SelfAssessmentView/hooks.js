import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

const useAssessmentViewHooks = () => {
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  return {
    submission: pageData?.submission,
    oraConfigData,
  };
};

export default useAssessmentViewHooks;
