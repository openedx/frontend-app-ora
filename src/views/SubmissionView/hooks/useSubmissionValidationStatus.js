import { useCallback } from 'react';
import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';

import {
  useSubmissionConfig,
} from 'hooks/app';

export const stateKeys = StrictDict({
  promptStatuses: 'promptStatuses',
  submissionTriggered: 'submissionTriggered',
  fileUploadIsRequired: 'fileUploadIsRequired',
});

const useSubmissionStatuses = (textResponses, uploadedFiles) => {
  const submissionConfig = useSubmissionConfig();

  // holds the status of each text field
  const [promptStatuses, setPromptStatuses] = useKeyedState(
    stateKeys.promptStatuses,
    textResponses.reduce((obj, val, i) => ({ ...obj, [i]: true }), {}),
  );
  const [submissionTriggered, setSubmissionTriggered] = useKeyedState(stateKeys.submissionTriggered, false);
  const [fileUploadIsRequired, setFileUploadIsRequired] = useKeyedState(stateKeys.fileUploadIsRequired, false);

  // this function is to be called when response submit button is clicked
  const validateBeforeConfirmation = useCallback(() => {
    setSubmissionTriggered(true);
    const retrievePromptsSubmissionStatus = textResponses.reduce((obj, val, i) => ({ ...obj, [i]: val }), {});
    let containsInvalidPrompts = false;
    if (submissionConfig.textResponseConfig.required) {
      for (let i = 0; i < textResponses.length; i++) {
        retrievePromptsSubmissionStatus[i] = !!textResponses[i];
      }
      containsInvalidPrompts = Object.values(retrievePromptsSubmissionStatus).some((val) => !val);
    }
    const calcFileUploadIsRequired = submissionConfig.fileResponseConfig.required && !uploadedFiles.length;
    setFileUploadIsRequired(calcFileUploadIsRequired);
    if (containsInvalidPrompts || (submissionConfig.fileResponseConfig.required && !uploadedFiles.length)) {
      setPromptStatuses(retrievePromptsSubmissionStatus);
      return false;
    }

    return true;
  }, [
    textResponses,
    uploadedFiles,
    submissionConfig,
    setPromptStatuses,
    setSubmissionTriggered,
    setFileUploadIsRequired,
  ]);

  return {
    validateBeforeConfirmation,
    setSubmissionTriggered,
    submissionTriggered,
    setPromptStatuses,
    promptStatuses,
    fileUploadIsRequired,
  };
};

export default useSubmissionStatuses;
