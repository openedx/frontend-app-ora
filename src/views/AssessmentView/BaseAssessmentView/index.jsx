import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Col, Row } from '@edx/paragon';

import { useViewStep } from 'hooks/routing';

import Assessment from 'components/Assessment';
import Instructions from 'components/Instructions';
import ModalActions from 'components/ModalActions';
import StatusAlert from 'components/StatusAlert';
import StepProgressIndicator from 'components/StepProgressIndicator';

import messages from '../messages';

import './BaseAssessmentView.scss';

const BaseAssessmentView = ({
  children,
}) => {
  const { formatMessage } = useIntl();
  const step = useViewStep();
  return (
    <div className="assessment-content-layout mr-auto ml-auto">
      <div className="content-wrapper">
        <StatusAlert />
        <Row className="flex-nowrap m-0 mb-3 position-relative">
          <div className="w-100">
            <h1 className="d-inline">{formatMessage(messages[step])}</h1>
            <StepProgressIndicator step={step} />
          </div>
        </Row>
        <Row className="flex-nowrap m-0 content-body">
          <Col className="p-0">
            <Instructions />
            {children}
            <ModalActions />
          </Col>
          <Col className="p-0 col-3 assessment-col">
            <Assessment />
          </Col>
        </Row>
      </div>
    </div>
  );
};
BaseAssessmentView.propTypes = {
  children: PropTypes.node.isRequired,
};

export default BaseAssessmentView;
