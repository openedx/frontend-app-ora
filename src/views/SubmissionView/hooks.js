import { useEffect, useReducer } from 'react';

import { useORAConfigData, usePageData } from 'data/services/lms/hooks/selectors';

import { submitResponse, saveResponse, uploadFiles } from 'data/services/lms/hooks/actions';
import { MutationStatus } from 'data/services/lms/constants';

const useSubmissionViewHooks = () => {
  const submitResponseMutation = submitResponse();
  const saveResponseMutation = saveResponse();
  const uploadFilesMutation = uploadFiles();
  const pageData = usePageData();
  const oraConfigData = useORAConfigData();

  const [submission, dispatchSubmission] = useReducer(
    (state, payload) => ({ ...state, isDirty: true, ...payload }),
    { ...pageData?.submission, isDirty: false },
  );

  useEffect(() => {
    // a workaround to update the submission state when the pageData changes
    if (pageData?.submission) {
      dispatchSubmission({ ...pageData.submission, isDirty: false });
    }
  }, [pageData?.submission]);

  const onTextResponseChange = (index) => (textResponse) => {
    dispatchSubmission({
      response: {
        ...submission.response,
        textResponses: [
          ...submission.response.textResponses.slice(0, index),
          textResponse,
          ...submission.response.textResponses.slice(index + 1),
        ],
      },
    });
  };

  const onFileUploaded = (args) => {
    const fileUploads = uploadFilesMutation.mutate(args);
    dispatchSubmission({ response: { ...submission.response, fileUploads } });
  };

  const submitResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    submitResponseMutation.mutate(submission);
  };

  const saveResponseHandler = () => {
    dispatchSubmission({ isDirty: false });
    saveResponseMutation.mutate(submission);
  };

  return {
    submitResponseHandler,
    submitResponseStatus: submitResponseMutation.status,
    saveResponseHandler,
    saveResponseStatus: saveResponseMutation.status,
    draftSaved: saveResponseMutation.status === MutationStatus.success && !submission.isDirty,
    pageData,
    oraConfigData,
    submission,
    dispatchSubmission,
    onTextResponseChange,
    onFileUploaded,
  };
};

export default useSubmissionViewHooks;
