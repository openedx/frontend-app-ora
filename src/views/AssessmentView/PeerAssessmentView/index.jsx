import React from 'react';

import {
  useIsORAConfigLoaded,
  usePrompts,
  useResponseData,
} from 'data/services/lms/hooks/selectors';

import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import FileUpload from 'components/FileUpload';
import BaseAssessmentView from '../BaseAssessmentView';

export const PeerAssessmentView = () => {
  const prompts = usePrompts();
  const response = useResponseData();
  if (!useIsORAConfigLoaded()) {
    return null;
  }
  return (
    <BaseAssessmentView submitAssessment={() => {}}>
      <div>
        {React.Children.toArray(
          prompts.map((prompt, index) => (
            <div>
              <Prompt prompt={prompt} />
              { response.textResponses && <TextResponse response={response.textResponses[index]} /> }
            </div>
          )),
        )}
        <FileUpload isReadOnly uploadedFiles={response.uploadedFiles} />
      </div>
    </BaseAssessmentView>
  );
};

export default PeerAssessmentView;
