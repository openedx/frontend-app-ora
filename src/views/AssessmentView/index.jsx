import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import Prompt from 'components/Prompt';
import TextResponse from 'components/TextResponse';
import FileUpload from 'components/FileUpload';
import { useViewStep } from 'hooks/routing';

import BaseAssessmentView from './BaseAssessmentView';
import useAssessmentData from './useAssessmentData';
import messages from './messages';

export const AssessmentView = () => {
  const { prompts, response, isLoaded } = useAssessmentData();
  const { formatMessage } = useIntl();
  const step = useViewStep();
  if (!isLoaded) {
    return null;
  }

  const textIsEmpty = !response?.textResponses?.length;
  const filesIsEmpty = !response?.uploadedFiles?.length;

  return (
    <BaseAssessmentView submitAssessment={() => {}}>
      <div>
        {React.Children.toArray(
          prompts.map((prompt, index) => (
            <div>
              <Prompt prompt={prompt} />
              {!textIsEmpty && (
                <>
                  <h3 className="m-1">{formatMessage(messages.responseMessages[step])}</h3>
                  <TextResponse response={response.textResponses[index]} />
                </>
              )}
            </div>
          )),
        )}
        {!filesIsEmpty && <FileUpload isReadOnly uploadedFiles={response.uploadedFiles} />}
      </div>
    </BaseAssessmentView>
  );
};

export default AssessmentView;
