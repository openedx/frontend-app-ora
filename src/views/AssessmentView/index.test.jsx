import { shallow } from '@edx/react-unit-test-utils';

import { useViewStep } from 'hooks/routing';

import { stepNames } from 'constants/index';
import useAssessmentData from './useAssessmentData';
import AssessmentView from './index';

jest.mock('hooks/routing', () => ({
  useViewStep: jest.fn(),
}));
jest.mock('components/Prompt', () => 'Prompt');
jest.mock('components/TextResponse', () => 'TextResponse');
jest.mock('components/FileUpload', () => 'FileUpload');
jest.mock('./useAssessmentData', () => jest.fn());

describe('<AssessmentView />', () => {
  useViewStep.mockReturnValue(stepNames.self);
  it('render empty on not loaded', () => {
    useAssessmentData.mockReturnValue({
      prompts: [],
      response: {},
      isLoaded: false,
    });
    const wrapper = shallow(<AssessmentView />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('render with prompts and text responses', () => {
    useAssessmentData.mockReturnValue({
      prompts: [
        { id: 1, prompt: 'prompt' },
        { id: 2, prompt: 'prompt' },
      ],
      response: {
        textResponses: [
          { id: 1, response: 'response' },
          { id: 2, response: 'response' },
        ],
      },
      isLoaded: true,
    });
    const wrapper = shallow(<AssessmentView />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Prompt').length).toBe(2);
    expect(wrapper.instance.findByType('TextResponse').length).toBe(2);
    expect(wrapper.instance.findByType('FileUpload').length).toBe(0);
  });

  it('render with files', () => {
    useAssessmentData.mockReturnValue({
      prompts: [],
      response: {
        uploadedFiles: [{ id: 1, name: 'file' }],
      },
      isLoaded: true,
    });
    const wrapper = shallow(<AssessmentView />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Prompt').length).toBe(0);
    expect(wrapper.instance.findByType('TextResponse').length).toBe(0);
    expect(wrapper.instance.findByType('FileUpload').length).toBe(1);
  });
});
