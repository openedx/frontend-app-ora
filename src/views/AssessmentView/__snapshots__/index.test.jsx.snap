// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssessmentView /> render empty on not loaded 1`] = `null`;

exports[`<AssessmentView /> render with files 1`] = `
<BaseAssessmentView
  submitAssessment={[Function]}
>
  <div>
    <FileUpload
      isReadOnly={true}
      uploadedFiles={
        Array [
          Object {
            "id": 1,
            "name": "file",
          },
        ]
      }
    />
  </div>
</BaseAssessmentView>
`;

exports[`<AssessmentView /> render with prompts and text responses 1`] = `
<BaseAssessmentView
  submitAssessment={[Function]}
>
  <div>
    <div
      key=".0"
    >
      <Prompt
        prompt={
          Object {
            "id": 1,
            "prompt": "prompt",
          }
        }
      />
      <Fragment>
        <h3
          className="m-1 py-2"
        >
          Your response
        </h3>
        <TextResponse
          response={
            Object {
              "id": 1,
              "response": "response",
            }
          }
        />
      </Fragment>
    </div>
    <div
      key=".1"
    >
      <Prompt
        prompt={
          Object {
            "id": 2,
            "prompt": "prompt",
          }
        }
      />
      <Fragment>
        <h3
          className="m-1 py-2"
        >
          Your response
        </h3>
        <TextResponse
          response={
            Object {
              "id": 2,
              "response": "response",
            }
          }
        />
      </Fragment>
    </div>
  </div>
</BaseAssessmentView>
`;
